### PR TITLE
Taggedmalloc

### PR DESCRIPTION
**Pull request by erikcorry**
Infrastructure to tag all allocations with a "process" or similar and later iterate over unfreed memory of a process.

### CODE REVIEW

File: components/heap/third_party/dartino/cmpctmalloc.c [hunk 1]
```diff
@@ -961,22 +1114,47 @@ IRAM_ATTR static void *page_alloc(cmpct_heap_t *heap, intptr_t pages)
...
+                heap->pages[i].tag = tag;
                 ASSERT(heap->pages[i].continued == 0);
                 for (int j = 1; j < pages; j++) {
                     heap->pages[i + j].in_use = 1;
                     heap->pages[i + j].continued = 1;
                 }
-                void *result = (void*)(heap->page_base + i * PAGE_SIZE);
+                void *result = heap->page_base + i * PAGE_SIZE;
                 for (int i = 0; i < pages << PAGE_SIZE_SHIFT; i += sizeof(int)) {
                     ((int*)(result))[i >> 2] = 0;
```
Comment (kasperl) [inline components/heap/third_party/dartino/cmpctmalloc.c]: `(int *)`

File: components/heap/third_party/dartino/cmpctmalloc.c [hunk 2]
```diff
@@ -202,8 +200,15 @@ typedef struct free_struct {
 typedef struct Page {
     uint32_t continued;  // Boolean - this is part 2 or more of a series of pages.
     uint32_t in_use;     // Boolean.
+    void *tag;
```
Comment (kasperl) [inline components/heap/third_party/dartino/cmpctmalloc.c]: Either a user-controlled tag or a heap reference, right? Add comment.

File: components/heap/third_party/dartino/cmpctmalloc.c [hunk 3]
```diff
@@ -129,7 +127,7 @@ static void page_free(cmpct_heap_t *heap, void *address, int pages_dummy);
 
 #define PAGE_SIZE_SHIFT 12
 #define PAGE_SIZE (1 << PAGE_SIZE_SHIFT)
-#define SMALL_ALLOCATION_LIMIT 4048
+#define SMALL_ALLOCATION_LIMIT 4032
```
[the remark this answers is not in the document]
Comment (erikcorry) [reply components/heap/third_party/dartino/cmpctmalloc.c]: Limited to the largest bucket size, and an assert added elsewhere to ensure this allocation size actually fits on a page.

File: components/heap/third_party/dartino/cmpctmalloc.c [hunk 4]
```diff
@@ -75,6 +66,9 @@ size_t multi_heap_free_size(cmpct_heap_t *heap)
 size_t multi_heap_minimum_free_size(cmpct_heap_t *heap)
     __attribute__((alias("cmpct_minimum_free_size_impl")));
 
+void multi_heap_iterate_tagged_memory_areas(cmpct_heap_t *heap, void *user_data, void *tag, tagged_memory_callback_t callback)
```
[the remark this answers is not in the document]
Comment (erikcorry) [reply components/heap/third_party/dartino/cmpctmalloc.c]: I just inherited this from the old implementation of malloc...

File: components/heap/third_party/dartino/cmpctmalloc.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (erikcorry) [reply components/heap/third_party/dartino/cmpctmalloc.c]: Also improved tests, and I think perhaps this didn't work before for some allocation sizes!

**Pull request merged**